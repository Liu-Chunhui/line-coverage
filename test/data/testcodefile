package identity

import (
	"context"
	errorz "errors"

	"github.com/yesino/pkg/errors"
	"github.com/yesino/pkg/errors/errcodes"
	"github.com/yesino/pkg/jwtauth"
	log "github.com/yesino/pkg/logging"
	"google.golang.org/grpc/codes"

	"github.com/yesino/pkg/sysmsg"
)

// Practically copied from yesino-cards' identity package

// Identity contains information extracted from an incoming JWT token about the id of the user calling
// an API.
type Identity struct {
	// A UUID that identifies the current user (A yesino person ID).
	UserID string
	// An identifier for the current person the user is acting as (1 user can map to many personas).
	PersonaID string
	// An identifier to know the auth type (FakeRock or yesino)
	Issuer string
}

// Get retrieves id information from JWT claims embedded in the context.
func Get(ctx context.Context) (*Identity, error) {
	claims, ok := jwtauth.GetClaimsFromContext(ctx)
	if !ok {
		log.Error(ctx, errorz.New("failed to extract JWT claims from context")).Send()
		return nil, errors.New(
			codes.Internal,
			sysmsg.AuthenticationGenericErrMsg,
			errors.NewErrorInfo(ctx, errcodes.CodeUnknown, "internal error"))
	}

	customer := claims.GetCustomer()
	if customer == nil {
		log.Error(ctx, errorz.New("failed to fetch customer details from JWT claims that is extracted from context")).Send()
		return nil, errors.New(
			codes.Internal,
			sysmsg.AuthenticationGenericErrMsg,
			errors.NewErrorInfo(ctx, errcodes.CodeUnknown, "internal error"))
	}

	if customer.GetSubject() == "" {
		log.Error(ctx, errorz.New("failed to fetch subject from JWT claims that is extracted from context")).Send()
		return nil, errors.New(
			codes.Internal,
			sysmsg.AuthenticationGenericErrMsg,
			errors.NewErrorInfo(ctx, errcodes.CodeUnknown, "internal error"))
	}

	if customer.GetPersona() == nil {
		log.Error(ctx, errorz.New("failed to fetch persona from JWT claims that is extracted from context")).Send()
		return nil, errors.New(
			codes.Internal,
			sysmsg.AuthenticationGenericErrMsg,
			errors.NewErrorInfo(ctx, errcodes.CodeUnknown, "internal error"))
	}

	return &Identity{
		customer.GetSubject(),
		customer.GetPersona().PersonaID,
		claims.GetBaseClaims().Issuer,
	}, nil
}
