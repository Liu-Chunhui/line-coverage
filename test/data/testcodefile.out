mode: atomic
github.com/Liu-Chunhui/line-coverage/test/data/testcodefile:14.80,16.16 2 1
github.com/Liu-Chunhui/line-coverage/test/data/testcodefile:19.2,24.6 4 1
github.com/Liu-Chunhui/line-coverage/test/data/testcodefile:44.2,44.19 1 1
github.com/Liu-Chunhui/line-coverage/test/data/testcodefile:16.16,18.3 1 0
github.com/Liu-Chunhui/line-coverage/test/data/testcodefile:24.6,26.17 2 60
github.com/Liu-Chunhui/line-coverage/test/data/testcodefile:34.3,35.17 2 59
github.com/Liu-Chunhui/line-coverage/test/data/testcodefile:39.3,39.17 1 59
github.com/Liu-Chunhui/line-coverage/test/data/testcodefile:26.17,27.21 1 1
github.com/Liu-Chunhui/line-coverage/test/data/testcodefile:31.4,31.19 1 0
github.com/Liu-Chunhui/line-coverage/test/data/testcodefile:27.21,28.10 1 1
github.com/Liu-Chunhui/line-coverage/test/data/testcodefile:35.17,37.4 1 0
github.com/Liu-Chunhui/line-coverage/test/data/testcodefile:39.17,41.4 1 59
github.com/Liu-Chunhui/line-coverage/test/data/testcodefile:47.62,48.29 1 63
github.com/Liu-Chunhui/line-coverage/test/data/testcodefile:59.2,59.19 1 60
github.com/Liu-Chunhui/line-coverage/test/data/testcodefile:48.29,50.17 2 6
github.com/Liu-Chunhui/line-coverage/test/data/testcodefile:54.3,54.12 1 6
github.com/Liu-Chunhui/line-coverage/test/data/testcodefile:50.17,52.4 1 0
github.com/Liu-Chunhui/line-coverage/test/data/testcodefile:54.12,56.4 1 3